tracer/test/Datadog.Trace.Tests/Util/Delegates/FuncInstrumentationTests.cs

```diff
@@ -0,0 +1,501 @@
+﻿// <copyright file="FuncInstrumentationTests.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.Util.Delegates;
+using FluentAssertions;
+using Xunit;
+
+namespace Datadog.Trace.Tests.Util.Delegates;
+
+#pragma warning disable SA1201
+
+public class FuncInstrumentationTests
+{
+    private delegate TReturn CustomFunc<TReturn>();
+
+    private delegate TReturn CustomFunc<T, TReturn>(T arg1);
+
+    private delegate TReturn CustomFunc<T, T2, TReturn>(T arg1, T2 arg2);
+
+    private delegate TReturn CustomFunc<T, T2, T3, TReturn>(T arg1, T2 arg2, T3 arg3);
+
+    private delegate TReturn CustomFunc<T, T2, T3, T4, TReturn>(T arg1, T2 arg2, T3 arg3, T4 arg4);
+
+    private delegate TReturn CustomFunc<T, T2, T3, T4, T5, TReturn>(T arg1, T2 arg2, T3 arg3, T4 arg4, T5 arg5);
+
+    [Fact]
+    public void Func0Test()
+    {
+        var callbacks = new Func0Callbacks();
+        Func<int> func = () =>
+        {
+            callbacks.Count.Value++;
+            return 42;
+        };
+        func = func.Instrument(callbacks);
+        func().Should().Be(43);
+        callbacks.Count.Value.Should().Be(3);
+
+        var value = 0;
+        CustomFunc<int> func2 = () =>
+        {
+            Interlocked.Increment(ref value).Should().Be(2);
+            return 42;
+        };
+        func2 = DelegateInstrumentation.Wrap(func2, new DelegateFunc0Callbacks(
+                                                 target =>
+                                                 {
+                                                     Interlocked.Increment(ref value).Should().Be(1);
+                                                     return null;
+                                                 },
+                                                 (target, returnValue, exception, state) =>
+                                                 {
+                                                     Interlocked.Increment(ref value).Should().Be(3);
+                                                     return ((int)returnValue) + 1;
+                                                 }));
```

Comment:
I wonder if we should actually include these `DelegateFunc0Callbacks` etc, or maybe just move them into the test project 🤔  Seems like we'd never want to use them in practice if they add extra allocations?

Reply:
I find this useful for prototyping or when we actually wants to instrument a delegate that runs only once, where the allocations are not so important.

Reply:
Fair enough, I was mostly thinking that we have enough complexity with this stuff already, having 2 ways to do things muddies the water, especially as it's not as good perf wise 🤷‍♂️ But I'm not passionate about it